File: frontend/src/api/schemas/school.ts
import { z } from 'zod'

/**
 * Schools 域走 snake_case wire（站点其它接口是 camelCase）。理由：
 * - 后端 sqlite 字段全 snake_case，复用 frontend/src/features/schools/types.ts
 *   原型零改动
 * - 后端用 SnakeModel 而非 CamelModel 对齐
 *
 * 字段顺序与 features/schools/types.ts 对齐；这里是 runtime 校验层。
 */

export const ReputationSchema = z.enum(['positive', 'neutral', 'negative', 'unknown'])
export const DegreeSchema = z.enum(['PhD', 'MS', 'Postdoc'])
export const RecruitFilterSchema = z.enum(['yes', 'no', 'unk'])
export const SortKeySchema = z.enum(['default', 'name', 'recruit', 'rep', 'updated'])
export const SortDirSchema = z.enum(['asc', 'desc'])

export const SchoolRefSchema = z.object({
  code: z.string(),
  name_cn: z.string(),
})

export const DeptRefSchema = z.object({
  code: z.string(),
  name_cn: z.string(),
})

export const QuotaSchema = z.object({
  year: z.number().int().nullable(),
  degree: DegreeSchema.nullable(),
  count: z.number().int().nullable(),
  confidence: z.number().nullable(),
  raw_text: z.string(),
  source_url: z.string().nullable(),
})

export const EvaluationSchema = z.object({
  source: z.string(),
  source_url: z.string().nullable(),
  content: z.string(),
  rating: z.number().nullable(),
  posted_at: z.string().nullable(),
})

export const TraceItemSchema = z.object({
  kind: z.string(),
  label: z.string(),
  detail: z.string(),
})

const AdvisorBaseShape = {
  id: z.number().int(),
  school: SchoolRefSchema,
  departments: z.array(DeptRefSchema),
  name_cn: z.string(),
  name_en: z.string().nullable(),
  title: z.string().nullable(),
  homepage: z.string(),
  source_url: z.string(),
  email: z.string().nullable(),
  email_obfuscated: z.boolean(),
  research_interests: z.array(z.string()),
  is_recruiting: z.boolean().nullable(),
  recruiting_confidence: z.number().nullable(),
  reputation_tag: ReputationSchema.nullable(),
  enriched_summary: z.string().nullable(),
  last_enriched_at: z.string().nullable(),
  note_count: z.number().int(),
  is_starred: z.boolean(),
}

export const AdvisorRowSchema = z.object(AdvisorBaseShape)
export type AdvisorRow = z.infer<typeof AdvisorRowSchema>

export const AdvisorDetailSchema = z.object({
  ...AdvisorBaseShape,
  phone: z.string().nullable(),
  photo_url: z.string().nullable(),
  bio_text: z.string().nullable(),
  quotas: z.array(QuotaSchema),
  evaluations: z.array(EvaluationSchema),
  trace: z.array(TraceItemSchema),
})
export type AdvisorDetail = z.infer<typeof AdvisorDetailSchema>

export const PaginatedAdvisorsSchema = z.object({
  items: z.array(AdvisorRowSchema),
  total: z.number().int().nonnegative(),
  page: z.number().int().positive(),
  page_size: z.number().int().positive(),
})
export type PaginatedAdvisors = z.infer<typeof PaginatedAdvisorsSchema>

export const SchoolMetaItemSchema = z.object({
  code: z.string(),
  name_cn: z.string(),
  name_en: z.string().nullable(),
  count: z.number().int().nonnegative(),
  departments: z.array(DeptRefSchema),
})

export const ManifestSchema = z.object({
  schema_version: z.number().int().nullable(),
  exported_at: z.string().nullable(),
  claw_version: z.string().nullable(),
  schools_sqlite_sha256: z.string().nullable(),
  schools_sqlite_bytes: z.number().int().nullable(),
  counts: z.record(z.string(), z.number()).nullable(),
})

export const SchoolsMetaSchema = z.object({
  schools: z.array(SchoolMetaItemSchema),
  titles: z.array(z.string()),
  manifest: ManifestSchema.nullable(),
})
export type SchoolsMeta = z.infer<typeof SchoolsMetaSchema>

export const ReloadResultSchema = z.object({
  ok: z.boolean(),
  manifest: ManifestSchema.nullable(),
})

export interface ListAdvisorsParams {
  school?: string[]
  dept?: string[]
  title?: string[]
  recruit?: Array<z.infer<typeof RecruitFilterSchema>>
  rep?: Array<z.infer<typeof ReputationSchema>>
  q?: string
  has_email?: boolean
  has_summary?: boolean
  sort_key?: z.infer<typeof SortKeySchema>
  sort_dir?: z.infer<typeof SortDirSchema>
  page?: number
  page_size?: number
}

File: backend/app/schemas/school.py
"""Pydantic schemas for the /schools/* domain.

Wire format here is snake_case — see app/schemas/_base.py::SnakeModel for
why this is the only place we break from site-wide camelCase. Field shapes
mirror frontend/src/features/schools/types.ts exactly so the frontend's
type definitions stay untouched.
"""
from __future__ import annotations

from typing import Literal

from app.schemas._base import SnakeModel, UtcDateTime

Reputation = Literal["positive", "neutral", "negative", "unknown"]
Degree = Literal["PhD", "MS", "Postdoc"]
RecruitFilter = Literal["yes", "no", "unk"]
SortKey = Literal["default", "name", "recruit", "rep", "updated"]
SortDir = Literal["asc", "desc"]


class SchoolRef(SnakeModel):
    code: str
    name_cn: str


class DeptRef(SnakeModel):
    code: str
    name_cn: str


class Quota(SnakeModel):
    year: int | None = None
    degree: Degree | None = None
    count: int | None = None
    confidence: float | None = None
    raw_text: str
    source_url: str | None = None


class Evaluation(SnakeModel):
    source: str
    source_url: str | None = None
    content: str
    rating: float | None = None
    posted_at: UtcDateTime | None = None


class TraceItem(SnakeModel):
    kind: str
    label: str
    detail: str


class AdvisorBase(SnakeModel):
    id: int
    school: SchoolRef
    departments: list[DeptRef]
    name_cn: str
    name_en: str | None = None
    title: str | None = None
    homepage: str
    source_url: str
    email: str | None = None
    email_obfuscated: bool
    research_interests: list[str]
    is_recruiting: bool | None = None
    recruiting_confidence: float | None = None
    reputation_tag: Reputation | None = None
    enriched_summary: str | None = None
    last_enriched_at: UtcDateTime | None = None
    # UGC overlay (filled by services/schools_overlay.py — placeholder for now)
    note_count: int = 0
    is_starred: bool = False


class AdvisorRow(AdvisorBase):
    """Slim row for the table view — no nested quotas/evaluations/trace."""


class AdvisorDetail(AdvisorBase):
    """Full detail for the drawer."""

    phone: str | None = None
    photo_url: str | None = None
    bio_text: str | None = None
    quotas: list[Quota]
    evaluations: list[Evaluation]
    trace: list[TraceItem]


class PaginatedAdvisors(SnakeModel):
    items: list[AdvisorRow]
    total: int
    page: int
    page_size: int


class SchoolMetaItem(SnakeModel):
    code: str
    name_cn: str
    name_en: str | None = None
    count: int
    departments: list[DeptRef]


class ManifestOut(SnakeModel):
    schema_version: int | None = None
    exported_at: str | None = None
    claw_version: str | None = None
    schools_sqlite_sha256: str | None = None
    schools_sqlite_bytes: int | None = None
    counts: dict[str, int] | None = None


class SchoolsMeta(SnakeModel):
    schools: list[SchoolMetaItem]
    titles: list[str]
    manifest: ManifestOut | None = None


class ReloadResult(SnakeModel):
    ok: bool
    manifest: ManifestOut | None = None
